### PR TITLE
wasm: include source location in runtime errors

### LINKED ISSUE #2862: Improve runtime error reporting for wasm compiled policies
Runtime errors do not include any source location information currently. This makes it difficult to pinpoint the source of the error in large policies. The planner and wasm backend should propagate source locations so that runtime errors can be identified more easily.

### PR DESCRIPTION
**Pull request by srenatus**
Note that due to the mapping of rego to imperative code, the errors
may look different than what you'd expect. For example, with this rego,

    package fr

    p {
      f(3)
    }

    f(x) = "y" {
      x > 1
    }

    f(x) = "z" {
      x > 2
    }

the runtime error is

    fun_rets.rego:12:3: var assignment conflict

Whereas `opa eval` gives you

    $ opa eval 'data.fr.p' -d fun_rets.rego --format=pretty
    1 error occurred: fun_rets.rego:11: eval_conflict_error: functions must not produce multiple outputs for same inputs

To get here, we're adding a new section to the static part of an ir.Policy: Files.
It's carrying the file name constants, and one special entry called `<query>`.

Most of the work in this change is ensuring that source location info is properly
carried over into the IR statements from AST nodes. There are new tests for the
planner "TestLocations", to assert a bunch of those mappings for different structures.
They attempt to not cover every statement, but a few select ones for each case.

The mappings works in an indirect way: there's a location stored in the planner's
state, `p.loc`, and it's used from `p.appendStmt`. For some constructions that bypass
this helper, there is an extra method, `p.appendStmtToBlock`, to carry over the location
from p.loc into the statements added that way.

Also:

* internal/ir: change location struct to include file and text

  This helps with debugging:

      | | | | *ir.BlockStmt BlockStmt (1 blocks)
      | | | | | *ir.Block Block (6 statements)
      | | | | | | *ir.MakeNumberRefStmt &{Index:1 Target:6 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
      | | | | | | *ir.MakeNumberRefStmt &{Index:2 Target:7 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
      | | | | | | *ir.GreaterThanStmt &{A:6 B:7 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
      | | | | | | *ir.MakeStringStmt &{Index:3 Target:8 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
      | | | | | | *ir.MakeStringStmt &{Index:4 Target:9 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
      | | | | | | *ir.ObjectInsertOnceStmt &{Key:8 Value:9 Object:5 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
      | | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}

  The fields are intentionally not exported, so they don't make their way
  into the WASM module.

* Fixes error return of ir.Walk()

  w.err would be ignored for the code path that walks leaves on the same level:

      w.walk(x.Static) // error set here
      w.walk(x.Plans)  // would be overwritten here
      w.walk(x.Funcs)

--------
Fixes #2862.

### CODE REVIEW

File: internal/compiler/wasm/wasm.go [hunk 1]
```diff
@@ -1264,3 +1276,25 @@ func getLowestFreeDataSegmentOffset(m *module.Module) (int32, error) {
 
 	return offset, nil
 }
+
+// runtimeErrorAbort uses the passed stmt's source location to build
+// the arguments for a call to opa_runtime_error(file, row, col, msg).
+// It returns the instructions that make up the function call with
+// arguments, followed by Unreachable.
+// TODO(sr): figure out if we can hit the loc == nil case and should
+// include an "<unknown>" or "<query>" pseudo-file of some sort.
```
Comment (srenatus) [inline internal/compiler/wasm/wasm.go:1285]: I haven't yet found an input that would trigger this case 🤔

File: internal/compiler/wasm/wasm.go [hunk 2]
```diff
@@ -1006,15 +1012,14 @@ func (c *Compiler) compileUpsert(local ir.Local, path []int, value ir.Local, ins
...
+			Instrs: append([]instruction.Instruction{
 				instruction.GetLocal{Index: ltemp},
 				instruction.Call{Index: c.function(opaValueType)},
 				instruction.I32Const{Value: opaTypeObject},
 				instruction.I32Eq{},
 				instruction.BrIf{Index: 0},
-				instruction.I32Const{Value: c.builtinStringAddr(errWithConflict)},
-				instruction.Call{Index: c.function(opaAbort)},
 			},
+			c.runtimeErrorAbort(loc, errWithConflict)...),
```
Comment (srenatus) [inline internal/compiler/wasm/wasm.go:1022]: ⚠️ This one didn't include `Unreachable` before, and now does. However, I suppose this code path will be revisited in #2922, too.

**Review by tsandall — COMMENTED** [review]
Looks great! This is one of the few cases where I think the struct embedding will be worth it. See my comment about maintaining the current location on the planner.

File: internal/planner/planner.go [hunk 3]
```diff
@@ -452,6 +457,8 @@ func (p *Planner) planQuery(q ast.Body, index int, iter planiter) error {
 
 // TODO(tsandall): improve errors to include location information.
 func (p *Planner) planExpr(e *ast.Expr, iter planiter) error {
+	p.setLocation(e.Location)
```
[the remark this answers is not in the document]
Comment (srenatus) [reply internal/planner/planner.go:460]: Yeah I was thinking about that problem. I hadn't been able to trigger it, but I'll have another look! Thanks.

[the remark this answers is not in the document]
Comment (tsandall) [reply internal/planner/planner.go:460]: Here's an example where the problem is observable.

The locations for statements in the function body are essentially correct (I'm assuming 0:1:1 is some empty/default value). The last three statements are at "0:6:4" (L4) which checks out given the test case:

```golang
		{
			note:    "functions",
			queries: []string{"data.test.f([1,x])"},
			modules: []string{`  # L1
				package test   # L2
				f([a, b]) {          # L3
					a = b       # L4
				}
			`},
		},
```

However, as you can see, the statements for the query have inherited the location info from the last expr in the function. This is because the call statements are planned AFTER the function definition has been planned.

```
$ go test ./internal/planner/ -v -run TestPlannerHelloWorld/functions
=== RUN   TestPlannerHelloWorld
=== RUN   TestPlannerHelloWorld/functions
*ir.Policy Policy
| *ir.Static Static (2 strings)
| | *ir.StringConst &{Value:x}
| | *ir.StringConst &{Value:1}
| *ir.Plans &{Plans:[Plan test (2 blocks)]}
| | *ir.Plan Plan test (2 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.MakeStringStmt <0:0:0>
| | | *ir.Block Block (10 statements)
| | | | *ir.MakeArrayStmt <0:6:4>
| | | | *ir.MakeNumberRefStmt <0:6:4>
| | | | *ir.ArrayAppendStmt <0:6:4>
| | | | *ir.ArrayAppendStmt <0:6:4>
| | | | *ir.CallStmt <0:6:4>
| | | | *ir.MakeBooleanStmt <0:6:4>
| | | | *ir.NotEqualStmt <0:6:4>
| | | | *ir.MakeObjectStmt <0:6:4>
| | | | *ir.ObjectInsertStmt <0:6:4>
| | | | *ir.ResultSetAdd <0:6:4>
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.test.f (3 params: [0 1 3], 2 blocks)
| | | *ir.Block Block (14 statements)
| | | | *ir.IsArrayStmt <0:1:1>
| | | | *ir.LenStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.EqualStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.EqualStmt <0:6:4>
| | | | *ir.MakeBooleanStmt <0:6:4>
| | | | *ir.AssignVarOnceStmt <0:6:4>
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt <0:0:0>
--- PASS: TestPlannerHelloWorld (0.00s)
    --- PASS: TestPlannerHelloWorld/functions (0.00s)
PASS
ok      github.com/open-policy-agent/opa/internal/planner       0.003s
```

[the remark this answers is not in the document]
Comment (srenatus) [reply internal/planner/planner.go:460]: > I'm assuming 0:1:1 is some empty/default value

Almost. It's the index into the files addresses array, row (line) and col. The first "file" we see typically is "<query>", so 0:1:1 would refer to the query. 

This indeed illustrates the problem nicely. My approach was silly, I had tried to trigger a runtime error instead of examining the IR 🤦‍♂️

I'm going to fiddle with the proper places to reset the location state.

[the remark this answers is not in the document]
Comment (srenatus) [reply internal/planner/planner.go:460]: Another 🤦 I had switched row/col in the `String()` method on location, it's meant to be 4:6, not 6:4.

[the remark this answers is not in the document]
Comment (srenatus) [reply internal/planner/planner.go:460]: Side show, but before https://github.com/open-policy-agent/opa/pull/2960/commits/8f1492b3896af6a552913b68f44f6b510b29238e we'd always thought our source locations were `<query>` because the filename set by `ast.MustParseModule` is `""`.

[the remark this answers is not in the document]
Comment (srenatus) [reply internal/planner/planner.go]: Resetting before calling `iter` (https://github.com/open-policy-agent/opa/pull/2960/commits/788ad9fb4e2bcbf424cef0c2978260f6f7dd16a4) got is to this:

```
*ir.Policy Policy
| *ir.Static Static (2 strings)
| | *ir.StringConst &{Value:x}
| | *ir.StringConst &{Value:1}
| *ir.Plans &{Plans:[Plan test (2 blocks)]}
| | *ir.Plan Plan test (2 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.MakeStringStmt <0:0:0>
| | | *ir.Block Block (10 statements)
| | | | *ir.MakeArrayStmt <0:1:1>
| | | | *ir.MakeNumberRefStmt <0:1:1>
| | | | *ir.ArrayAppendStmt <0:1:1>
| | | | *ir.ArrayAppendStmt <0:1:1>
| | | | *ir.CallStmt <0:1:1>
| | | | *ir.MakeBooleanStmt <0:1:1>
| | | | *ir.NotEqualStmt <0:1:1>
| | | | *ir.MakeObjectStmt <0:0:0>
| | | | *ir.ObjectInsertStmt <0:0:0>
| | | | *ir.ResultSetAdd <0:0:0>
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.test.f (3 params: [0 1 3], 2 blocks)
| | | *ir.Block Block (14 statements)
| | | | *ir.IsArrayStmt <0:1:1>
| | | | *ir.LenStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.EqualStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.EqualStmt <1:4:6>
| | | | *ir.MakeBooleanStmt <0:1:1>
| | | | *ir.AssignVarOnceStmt <0:1:1>
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt <0:0:0>
```
Note that the row/cols are switched compared to before. Also, we not get a true zero value, `<0:0:0>`.

**Comment by srenatus** [comment]
(Just scribbles)

These cause trouble:
```
		{
			note:    "functions",
			queries: []string{"data.test.f([1, x, 2==2])"},
			modules: []string{`
				package test # L2
				f([a, b, true]) {  # L3
					a = b    # L4
				}
			`},
		},
```
`planner_test.go:246: call term not implemented`

```
		{
			note:    "functions",
			queries: []string{"data.test.f([1, x, true])"},
			modules: []string{`
				package test # L2
				f([a, b, 2==2]) {  # L3
					a = b    # L4
				}
			`},
		},
```
`planner_test.go:246: not implemented: unifyLocal(7, equal(2, 2))`

**Comment by srenatus** [comment]
We're not there yet:
```
*ir.Policy Policy
| *ir.Static Static (5 strings)
| | *ir.StringConst &{Value:x}
| | *ir.StringConst &{Value:1}
| | *ir.StringConst &{Value:0}
| | *ir.StringConst &{Value:a}
| | *ir.StringConst &{Value:b}
| *ir.Plans &{Plans:[Plan test (2 blocks)]}
| | *ir.Plan Plan test (2 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.MakeStringStmt &{Index:0 Target:2 Location:{Index:0 Col:0 Row:0 file: text:}}
| | | *ir.Block Block (5 statements)
| | | | *ir.CallStmt &{Func:g0.data.test.a Args:[0 1] Result:3 Location:{Index:1 Col:1 Row:1 file:<query> text:data.test.a = x}}
| | | | *ir.AssignVarStmt &{Source:3 Target:4 Location:{Index:1 Col:1 Row:1 file:<query> text:data.test.a = x}}
| | | | *ir.MakeObjectStmt &{Target:5 Location:{Index:1 Col:1 Row:1 file:<query> text:data.test.a = x}}
| | | | *ir.ObjectInsertStmt &{Key:2 Value:4 Object:5 Location:{Index:1 Col:1 Row:1 file:<query> text:data.test.a = x}}
| | | | *ir.ResultSetAdd &{Value:5 Location:{Index:1 Col:1 Row:1 file:<query> text:data.test.a = x}}
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.test.a (2 params: [0 1], 2 blocks)
| | | *ir.Block Block (6 statements)
| | | | *ir.MakeBooleanStmt &{Value:true Target:3 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.NotEqualStmt &{A:3 B:4 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.MakeObjectStmt &{Target:5 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.BlockStmt BlockStmt (1 blocks)
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.MakeNumberRefStmt &{Index:1 Target:6 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.MakeNumberRefStmt &{Index:2 Target:7 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.GreaterThanStmt &{A:6 B:7 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.MakeStringStmt &{Index:3 Target:8 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.MakeStringStmt &{Index:4 Target:9 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.ObjectInsertOnceStmt &{Key:8 Value:9 Object:5 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt &{Source:2 Location:{Index:0 Col:0 Row:0 file: text:}}
```

Notably, 
```
| | | | | | *ir.MakeStringStmt &{Index:3 Target:8 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.MakeStringStmt &{Index:4 Target:9 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
| | | | | | *ir.ObjectInsertOnceStmt &{Key:8 Value:9 Object:5 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
```
have nothing to do with `1 > 0` but should point to the expression `{ "a": "b" | 1 > 0 }`. 
```
| | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1 > 0}}
```
should point to `a = { ... }`.

**Comment by srenatus** [comment]
💭 Taking the location from an `ast.Term`, too, isn't too bad either, perhaps:
```
| | *ir.Func g0.data.test.a (2 params: [0 1], 2 blocks)
| | | *ir.Block Block (6 statements)
| | | | *ir.MakeBooleanStmt &{Value:true Target:3 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.MakeBooleanStmt &{Value:false Target:4 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.NotEqualStmt &{A:3 B:4 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.MakeObjectStmt &{Target:5 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | | *ir.BlockStmt BlockStmt (1 blocks)
| | | | | *ir.Block Block (6 statements)
| | | | | | *ir.MakeNumberRefStmt &{Index:1 Target:6 Location:{Index:0 Col:23 Row:4 file:module-0.rego text:1}}
| | | | | | *ir.MakeNumberRefStmt &{Index:2 Target:7 Location:{Index:0 Col:27 Row:4 file:module-0.rego text:0}}
| | | | | | *ir.GreaterThanStmt &{A:6 B:7 Location:{Index:0 Col:27 Row:4 file:module-0.rego text:0}}
| | | | | | *ir.MakeStringStmt &{Index:3 Target:8 Location:{Index:0 Col:11 Row:4 file:module-0.rego text:"a"}}
| | | | | | *ir.MakeStringStmt &{Index:4 Target:9 Location:{Index:0 Col:16 Row:4 file:module-0.rego text:"b"}}
| | | | | | *ir.ObjectInsertOnceStmt &{Key:8 Value:9 Object:5 Location:{Index:0 Col:16 Row:4 file:module-0.rego text:"b"}}
| | | | *ir.AssignVarOnceStmt &{Target:2 Source:5 Location:{Index:0 Col:9 Row:4 file:module-0.rego text:{ "a": "b" |  1 > 0 }}}
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt &{Source:2 Location:{Index:0 Col:0 Row:0 file: text:}}
```

Having MakeNumberRefStmt and MakeStringStmt reference the exact term they're dealing seems nice.

**Comment by srenatus** [comment]
☝️ Sorry, I've been commenting on my local WIP state, not correlated that last comment to any comment pushed.

**Comment by tsandall** [comment]
The errors mentioned in [this](https://github.com/open-policy-agent/opa/pull/2960#issuecomment-738668247) comment are expected because the planner test doesn't run the ast.Compiler on the queries/modules. As a result, all of the rewriting done by ast.Compiler is missing (e.g., expressions like `f(x)[_]` haven't been turned into `f(x, $0); $0[_]` and so on.)

**Comment by srenatus** [comment]
TODOs:
- [ ] examine, and fix the e2e test failures in wasm runtime error assertions

File: internal/planner/planner.go [hunk 4]
```diff
@@ -248,6 +248,7 @@ func (p *Planner) planRules(rules []*ast.Rule) (string, error) {
 				err := p.planQuery(rule.Body, 0, func() error {
 
 					// Run planner on the result.
+					p.loc = rule.Head.Loc() // reset location
```
Comment (tsandall) [inline internal/planner/planner.go:251]: This is somewhat minor, but I wonder if we can fix `planQuery` so that when this inner function is called, the location set above (at L224) is intact. I think it would be best if callers can assume the location is not overwritten.

File: internal/planner/planner.go [hunk 5]
```diff
@@ -299,15 +300,24 @@ func (p *Planner) planRules(rules []*ast.Rule) (string, error) {
...
-			},
-		})
+		// Set the location for the default rule head.
+		p.loc = defaultRule.Loc()
+		// NOTE(sr) for `default p = 1`,
+		// defaultRule.Loc() is `default`,
+		// defaultRule.Head.Loc() is `p = 1`.
+		// I'd prefer we'd have `default p = 1`; but since the mapped
+		// locations are col/row only (no extent), pointing to `default`
+		// seems preferable.
```
Comment (tsandall) [inline internal/planner/planner.go:310]: SGTM

File: internal/planner/planner.go [hunk 6]
```diff
@@ -322,14 +332,10 @@ func (p *Planner) planRules(rules []*ast.Rule) (string, error) {
 		}
 	}
 
+	p.loc = ploc // location of the ReturnLocalStmt is its call site -- TODO(sr): good choice?
```
Comment (tsandall) [inline internal/planner/planner.go:335]: Nit: I'd prefer if the location was associated with the rule (either the head or the last line in the body). Otherwise we'll get different location info depending on the order we plan the rules in (I think--because if you have multiple calls to the same rule we only plan the rule once...)

File: internal/planner/planner.go [hunk 7]
```diff
@@ -299,15 +300,24 @@ func (p *Planner) planRules(rules []*ast.Rule) (string, error) {
...
+		// locations are col/row only (no extent), pointing to `default`
+		// seems preferable.
+
+		fn.Blocks = append(fn.Blocks, p.blockWithStmt(&ir.IsUndefinedStmt{Source: fn.Return}))
+
+		p.loc = defaultRule.Head.Loc()
 
 		p.curr = fn.Blocks[len(fn.Blocks)-1]
 
+		// TODO(sr): default rules can have bodies?!
```
Comment (tsandall) [inline internal/planner/planner.go:318]: For better or worse, they're just represented as normal rules. The parser will generate a body that never fails (`true`). This could be useful in the future we if we wanted to relax some of the restrictions around `default`...

```
package x

default p = q   # q is always defined so this is fine however today it'll be rejected

default q = 7

File: internal/planner/planner.go [hunk 8]
```diff
@@ -375,6 +381,8 @@ func (p *Planner) planQueries() error {
 		for _, qv := range qvs.Sorted() {
 			qv = rewrittenVar(qs.RewrittenVars, qv)
 			if !qv.IsGenerated() && !qv.IsWildcard() {
+				// NOTE(sr): We have no location for these: they could appear in multiple
```
Comment (tsandall) [inline internal/planner/planner.go:384]: This is probably fine. These statements are unlikely to be interesting in terms of error messaging, profiling, etc.

File: internal/planner/planner.go [hunk 9]
```diff
@@ -445,13 +453,21 @@ func (p *Planner) planQuery(q ast.Body, index int, iter planiter) error {
...
+	}()
+
 	return p.planExpr(q[index], func() error {
 		return p.planQuery(q, index+1, iter)
 	})
 }
 
 // TODO(tsandall): improve errors to include location information.
 func (p *Planner) planExpr(e *ast.Expr, iter planiter) error {
+	// p.loc = e.Location
```
Comment (tsandall) [inline internal/planner/planner.go:469]: nit: delete

File: internal/planner/planner.go [hunk 10]
```diff
@@ -683,6 +700,7 @@ func (p *Planner) planExprCall(e *ast.Expr, iter planiter) error {
 		})
 	case ast.GreaterThan.Name:
 		return p.planBinaryExpr(e, func(a, b ir.Local) error {
+			p.loc = ploc // XXX: exemplary, all of those above and below would need this
```
Comment (tsandall) [inline internal/planner/planner.go:703]: Is this related to providing term-level locations? I'm just wondering why we need this.

File: internal/planner/planner.go [hunk 11]
```diff
@@ -1360,15 +1380,18 @@ func (p *Planner) planComprehension(body ast.Body, closureIter planiter, target
 
 	prev := p.curr
 	p.curr = &ir.Block{}
+	ploc := p.loc
 
 	if err := p.planQuery(body, 0, func() error {
+		p.loc = ploc
```
Comment (tsandall) [inline internal/planner/planner.go:1386]: I suspect if we make the change I mentioned above, this will go away?

**Review by tsandall — COMMENTED** [review]
This PR is progressing very nicely. Left a few comments w/ thoughts on the planner location updates.

File: go.mod [hunk 12]
```diff
@@ -9,6 +9,7 @@ require (
 	github.com/ghodss/yaml v1.0.0
 	github.com/gobwas/glob v0.2.3
 	github.com/golang/protobuf v1.4.3 // indirect
+	github.com/google/go-cmp v0.5.4
```
[the remark this answers is not in the document]
Comment (srenatus) [reply go.mod:12]: leftover, will cleanup.

File: internal/planner/planner.go (hunk 10 above)
Comment (srenatus) [reply internal/planner/planner.go:703]: Yeah will cleanup. 🙈

File: internal/planner/planner.go (hunk 5 above)
Comment (srenatus) [reply internal/planner/planner.go:310]: 😄 So, overnight, I've kind of convinced myself that this was the wrong choice. I think what matters is the `p = 1`, much more so than the `default`, when it comes to potential errors related to this.

File: internal/planner/planner.go (hunk 7 above)
Comment (srenatus) [reply internal/planner/planner.go:317]: 💡 Ah, I see. Thanks!

File: internal/planner/planner.go (hunk 4 above)
Comment (srenatus) [reply internal/planner/planner.go]: :+1: done

File: internal/planner/planner.go (hunk 6 above)
Comment (srenatus) [reply internal/planner/planner.go:334]: Good point.

> either the head or the last line in the body

We'll now record the first rule's location, and use that. (I think it makes more sense than the last rule, or default rule if there is one.) WDYT?

**Comment by srenatus** [comment]
ℹ️ Covered a few more cases where statements were added without location info or with the wrong one.

File: internal/planner/planner.go (hunk 6 above)
Comment (tsandall) [reply internal/planner/planner.go]: The first rule's location makes the most sense. I hadn't thought it through completely but I now realize the last line wouldn't make sense because this will apply to all types of rules.

File: internal/planner/planner.go (hunk 5 above)
Comment (tsandall) [reply internal/planner/planner.go]: I'm fine with either--in practice, I think the row will be the important bit and the answer will be the same.

**Comment by srenatus** [comment]
This is ready, let's iterate from here. I'll squash all those commits tomorrow.

**Comment by srenatus** [comment]
✔️ squashed and rebased.

File: internal/planner/planner_test.go [hunk 13]
```diff
@@ -19,13 +21,12 @@ func TestPlannerHelloWorld(t *testing.T) {
 	// test/wasm/ directory that are specified in YAML, compiled into Wasm, and
 	// executed inside of a node program. For the time being, the planner is
 	// simple enough that exhaustive unit testing is not as valuable as
-	// end-to-end testing. These tests provide a quick sanity check that the
+	// end-to-end testing. These tests provide a quick consistency check that the
```
Comment (tsandall) [inline internal/planner/planner_test.go:24]: Thanks for changing this comment here and below! Sanity check isn't very clear and is somewhat ableist. 👍

File: internal/planner/planner.go [hunk 14]
```diff
@@ -445,13 +454,20 @@ func (p *Planner) planQuery(q ast.Body, index int, iter planiter) error {
...
+	p.loc = q[index].Loc()
+
 	return p.planExpr(q[index], func() error {
-		return p.planQuery(q, index+1, iter)
+		return p.planQuery(q, index+1, func() error {
+			p.loc = old
+			return iter()
+		})
 	})
 }
```
Comment (tsandall) [inline internal/planner/planner.go:466]: I think we need to reset the location before the final return to ensure that the caller's location is not lost. Similarly, the inner-most iterator should do the same.

```golang
old := p.loc
p.loc := q[index].Loc()

err := p.planExpr(q[index], func() error {
   return p.planQuery(q, index+1, func() error {
      curr := p.loc
      p.loc = old
      err := iter()
      p.loc = curr
      return err
   })
})

p.loc = old
return err
```

E.g., when the iterator function on L460 returns back into `planExpr`, the location should be set to the location of that expression. If there was iteration happening inside `planExpr` where the iterator function L460 was called multiple times, I think we could observe this.

File: internal/planner/planner.go [hunk 15]
```diff
@@ -1852,8 +1861,37 @@ func (p *Planner) getStringConst(s string) int {
...
+		}
+		s.SetLocation(p.getFileConst(str), p.loc.Row, p.loc.Col, str, string(p.loc.Text))
+	}
+	b.Stmts = append(b.Stmts, s)
+}
+
+func (p *Planner) blockWithStmt(s ir.Stmt) *ir.Block {
+	b := &ir.Block{}
+	p.appendStmtToBlock(s, b)
+	return b
```
Comment (tsandall) [inline internal/planner/planner.go:1894]: Nice!

**Review by tsandall — DISMISSED** [review]
One minor comment on the planQuery change. Other than that, LGTM.

File: internal/planner/planner.go (hunk 14 above)
Comment (srenatus) [reply internal/planner/planner.go]: Thanks! It fixes one test's location that was odd, too. 🎉

File: internal/planner/planner_test.go (hunk 13 above)
Comment (srenatus) [reply internal/planner/planner_test.go:24]: Yeah, I've taken this opportunity to sneaked this in. 😄

**Review by tsandall — APPROVED** [review]
LGTM. Once that failing test is fixed let's merge this!

**Comment by srenatus** [comment]
✔️ Test fixed, commits squashed. If you find a moment, I'd appreciate this getting merged 😄
